Lrp follow-up fixes

aditighag opened the pull request:
Check commit description.

Addresses some of the follow-ups.

Fixes: #13294

aditighag committed:
pkg/redirectpolicy Improve error logging/function names

- Improve error logging in getting pod metadata to
handle accessing null fields
- Rename function names to better reflect their implementation
- Change lock type to Mutex since we mainly have write paths

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/redirectpolicy Refactor frontend mapped backends

Store podID along with addresses for the
frontend mapped backends so that updates/deletes for
the backend pods can be handled effectively when
there are multiple backend pods for a frontend.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
Documentation Update gsg with motivations and run-time behavior

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service Prevent local-redirect svc from overriding other types

A local-redirect service can only override itself or
a clusterIP service type.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/watcher Add LocalRedirectPolicy group to k8s sync list

K8s service sync is triggered when cilium agent is
restarted in order to keep track of stale services that
were removed when cilium agent was down.
Ensure LocalRedirectPolicy updates are processed
before the sync is triggered so that active services
are marked accordingly, and won't be removed during
the sync.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on pkg/service/service_test.go:
```diff
@@ -621,4 +622,31 @@ func (m *ManagerTestSuite) TestLocalRedirectServiceOverride(c *C) {
+
+	// Insert the service entry of type Local Redirect.
+	p2.Type = lb.SVCTypeLocalRedirect
+	created, id, err = m.svc.UpsertService(p2)
```
I've fixed the failure flagged by GH actions.

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -250,7 +250,11 @@ func (rpm *Manager) OnUpdatePodLocked(pod *slimcorev1.Pod) {
-	podData := rpm.getPodMetadata(pod, podIPs)
+	var podData *podMetadata
+	if podData, err = rpm.getPodMetadata(pod, podIPs); err != nil {
+		log.Debugf("failed to get valid pod metadata %s", err)
```
`log.WithError(err).Debug(...`

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -501,7 +498,11 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
+	var (
+		retPods []*podMetadata
+		podData *podMetadata
+		e       error
```
`s/e/err`

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -512,11 +513,14 @@ func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig, podStore cache.Stor
-		podInfo := rpm.getPodMetadata(pod, podIPs)
-		if !config.policyConfigSelectsPod(podInfo) {
+		if podData, e = rpm.getPodMetadata(pod, podIPs); e != nil {
+			log.Debugf("failed to get valid pod metadata %s", e)
```
`log.WithError(err).Debug`

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -396,23 +389,15 @@ func (rpm *Manager) updateConfigSvcFrontend(config *LRPConfig, frontends ...*fro
+			// Remove the pod from the frontend's backends slice.
+			if be.podID == podID {
+				fe.podBackends[i] = fe.podBackends[len(fe.podBackends)-1]
+				fe.podBackends = fe.podBackends[:len(fe.podBackends)-1]
```
I would have found a little bit surprising the order of backends to be changed if I was deleting them. I would suggest to add this caveat in the function description or keep the order on deletion.

Also, needs unit tests.

aditighag replied:
The load balancing framework does round-robin during service translation anyway so the order is not super critical, is it? I'll keep the order same nevertheless.

joestringer replied:
Probably @brb would be the one to give you a better idea of whether ordering matters. I wouldn't be surprised if it plays into the "service stickiness" implementation, but I don't actually know about this off the top of my head.

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -672,22 +662,43 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
+	for _, be := range frontendMapping.podBackends {
+		if be.podID != podID {
+			newFePods = append(newFePods, be)
+		}
```
can you add a comment describing what this logic is doing? It's not clear to me.

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -672,22 +662,43 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
+		for _, poID := range podPolicies {
+			// Existing pod policy update
+			if poID == config.id {
+				newPodPolicy = false
```
```
newPodPolicy = false
break
```

aanm commented on pkg/redirectpolicy/redirectpolicy.go:
```diff
@@ -81,15 +81,18 @@ type LRPConfig struct {
-type backend = loadbalancer.L3n4Addr
+type backend struct {
+	loadbalancer.L3n4Addr
+	podID podID
```
Add a comment describing what is this `podID`

brb commented on pkg/k8s/watchers/cilium_local_redirect_policy.go:
```diff
@@ -72,7 +72,7 @@ func (k *K8sWatcher) ciliumLocalRedirectPolicyInit(ciliumLRPClient *k8s.K8sCiliu
 
 	go lrpController.Run(wait.NeverStop)
-	k.k8sAPIGroups.addAPI(k8sAPIGroupCiliumClusterwideNetworkPolicyV2)
+	k.k8sAPIGroups.addAPI(k8sAPIGroupCiliumLocalRedirectPolicyV2)
```
Q: why don't we sync on `k8sAPIGroupCiliumClusterwideNetworkPolicyV2`, and instead we sync on `k8sAPIGroupCiliumLocalRedirectPolicyV2`?

aditighag replied:
Per the commit description, we want LRP updates to be processed in order to keep track of active local-redirect services before k8s service sync is called.

joestringer commented on Documentation/gettingstarted/local-redirect-policy.rst:
```diff
@@ -13,24 +13,33 @@ Local Redirect Policy (beta)
+<https://kubernetes.io/docs/reference/access-authn-authz/rbac/>`_.
+
+When policies are applied, application developers can not escape redirection
+enforced on pod traffic that matches the policy configurations.
```
nit: This makes assumptions about the RBAC configuration, but it's stated as a general truth. I think it would be more generally correct if we simply state that the traffic is redirected when this is configured, and describe that control over such configurations is controlled by the RBAC configuration.

joestringer commented on Documentation/gettingstarted/local-redirect-policy.rst:
```diff
@@ -154,24 +176,54 @@ Verify that the traffic was redirected to the ``lrp-pod`` that was deployed.
 
 This type of configuration is specified using Kubernetes service name and namespace
-for which traffic needs to be redirected.
+for which traffic needs to be redirected. The service needs to be of type ``clusterIP``.
```
nit: 
```suggestion
for which traffic needs to be redirected. The service must be of type ``clusterIP``.
```

joestringer commented on Documentation/gettingstarted/local-redirect-policy.rst:
```diff
@@ -189,27 +241,23 @@ Verify that the custom resource is created.
-    $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes-local-redirect/k8s-svc.yaml
-
-Verify that the service is created.
+Verify that entry Cilium's eBPF kube-proxy replacement updated the previously
```
nit, simplification:

```suggestion
Verify that entry Cilium updated the previously
```

joestringer replied:
The words "previously created" are probably not necessary either, this section is all talking about one specific service so I don't think this is ambiguous.

joestringer commented on Documentation/gettingstarted/local-redirect-policy.rst:
```diff
@@ -220,14 +268,37 @@ Invoke a curl command from the client pod to the Cluster IP address and port of
+    01:36:24.608867 IP 10.16.215.55.60876 > 10.16.70.187.80: Flags [P.], seq 1:96, ack 1, win 221, options [nop,nop,TS val 2541637677 ecr 2962246962], length 95: HTTP: HEAD /index.html HTTP/1.1
+    01:36:24.608876 IP 10.16.70.187.80 > 10.16.215.55.60876: Flags [.], ack 96, win 219, options [nop,nop,TS val 2962246962 ecr 2541637677], length 0
+    01:36:24.609007 IP 10.16.70.187.80 > 10.16.215.55.60876: Flags [P.], seq 1:239, ack 96, win 219, options [nop,nop,TS val 2962246962 ecr 2541637677], length 238: HTTP: HTTP/1.1 200 OK
+    01:36:24.609052 IP 10.16.215.55.60876 > 10.16.70.187.80: Flags [.], ack 239, win 229, options [nop,nop,TS val 2541637677 ecr 2962246962], length 0
```
I still feel like telling users to run tcpdump to verify the behaviour is very detached from the use case of LRP. For development and debugging, it makes sense. To demonstrate the value of the feature, it makes less sense. Tcpdumping the traffic to verify the redirect is not an end goal, it just proves that the functionality works (which should be safe to assume). From here, the user still needs to interpret this for their use case to understand the gap between this point in the GSG and the redirect behaviour they want, such as achieving node-local DNS.

I'm OK with this for now given that it's better to document and demonstrate rather than have no documentation and be unclear how to validate the behaviour, but I think we can do a better job of demonstrating the value of the feature with a real-world use case such as deploying node-local DNS.

aditighag replied:
Agreed, demonstrating real-world use case is great. I'll keep the simple example for now. I'll also also add references to the kiam/node-local-dns example yamls and the corresponding issues with more details when both the cases are completely fixed.

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -66,7 +66,7 @@ type Manager struct {
 
 	// Mutex to protect against concurrent access to the maps
-	mutex lock.RWMutex
+	mutex lock.Mutex
```
I'm OK with this if the access is not going to take advantage of many readers / few writers. But I thought I saw some other PR using `RLock()` so let's make sure to track this to resolve code conflicts before merging, so as not to break the tree. (/cc @Weil0ng )

joestringer replied:
(To be clear, this alone should not block the PR, we should just be careful about re-running CI between merging this code and any other PR that may be attempting to use `RLock()`, regardless of which order the PRs go in).

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -250,7 +250,11 @@ func (rpm *Manager) OnUpdatePodLocked(pod *slimcorev1.Pod) {
+	var podData *podMetadata
+	if podData, err = rpm.getPodMetadata(pod, podIPs); err != nil {
+		log.Debugf("failed to get valid pod metadata %s", err)
+		return
```
The changes in this commit around `getPodMetadata()` make me a bit uneasy, specifically because:
1. Debug messages will never be printed in prod environments, so if this error condition is ever hit then the user has no way to confirm why the LRP doesn't work.
1. The commit message doesn't call out the intent of catching and ignoring these error messages, so there's no way for future readers in git to reason about why this code was added and whether it was intentional or a mistake.
1. The commit shares other largely cosmetic changes like making functions private, `upsertConfig` -> `processConfig` rename, the mutex change, etc.

IIRC the original concern with this code was the possibility that bad user input could trigger a segfault. Any fix to an issue like this should be placed in a separate commit to cosmetic changes.  It seems like you did a good job of this for the other commits, but this one didn't seem so obvious to split out.

As an aside, I think we should strive to avoid commit messages like "Address review feedback". The commit title should reflect the intent of the change and how it will improve the code, whether it's the functionality, structure, readability, etc. When reading a git log for some code, "address feedback" provides zero context. I assume that all code that goes in should address feedback. I'd much rather read in a git log why this specific line was inserted - was it fixing a bug or refactoring or improving readability? Sometimes we find bugs purely because there is a discrepancy between the intent expressed in the message and the change made in the code.

aditighag replied:
>Debug messages will never be printed in prod environments, so if this error condition is ever hit then the user has no way to confirm why the LRP doesn't work.

The checks done by `getPodMetadata` (and similar existing code in cilium [1] ) are more of back-up? These checks are already done by Kubernetes. But I'm fine with changing the log level for the error cases to error instead of debug.

```
kubectl apply -f examples/kubernetes-local-redirect/backend-pod.yaml
The Pod "lrp-pod" is invalid: spec.containers[0].ports[1].protocol: Unsupported value: "uddp": supported values: "SCTP", "TCP", "UDP"
```

Re: generic commit title: noted. 

[1] https://github.com/cilium/cilium/blob/ada53a91a98618d24dad2abda97dd9c60923003c/pkg/k8s/watchers/pod.go#L637

joestringer replied:
Looking at `updatePodHostData()` and `getPodMetadata()`, the checks look almost identical. Should `updatePodHostData()` be guaranteed to be called before hitting `getPodMetadata()`? Do you know of any way this would not be the case?

Overall, if we validate the inputs earlier on and prevent hitting these cases, then there's no point in doing additional validation in `getPodMetadata()` and this whole error checking question becomes moot.

aditighag replied:
The checks might be required for the code path where we fetch pods from the `podStore`, in which case `updatePodHostData` isn't called.

joestringer replied:
OK, makes sense. The changes here LGTM.

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -396,23 +389,15 @@ func (rpm *Manager) updateConfigSvcFrontend(config *LRPConfig, frontends ...*fro
+			// Remove the pod from the frontend's backends slice, keeping the
+			// order same.
+			if be.podID == podID {
+				fe.podBackends = append(fe.podBackends[:i], fe.podBackends[i+1:]...)
```
Are we unit-testing this?

I'm not sure the slices are guaranteed to behave the way you intend here, since this code is attempting to modify `fe.podBackends` while iterating it.

https://play.golang.org/p/o3BoJobIK9D

aditighag replied:
I did test this code by adding multiple backend pods, but I didn't have multiple pod <ip, port> tuples. I'll fix it.

aanm replied:
Please also add unit tests.

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -672,22 +661,53 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
+func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
+	newFePods := make([]backend, 0, len(frontendMapping.podBackends))
+	// Update the frontend mapped backends slice, keeping the order same.
+	// TODO (aditi) unit test case
```
+1, do you intend to address that in this PR?

Having tests in place before refactoring is a great way to provide reassurance that things are still working like they were before.

aditighag replied:
> If I'm reading it right, if the pod has multiple backends then we'll issue multiple service updates but we should be able to just calculate all backends for a service at once and then update for all known backends.

We keep a map of frontend and its mapped backends. A frontend may have multiple backend pods, each with multiple backends (IP+port tuples). Whenever there is an update for a backend pod, we update all the frontend mappings that are mapped to this backend pod. Before we can override the current service entry for the frontend, we re-compute the backends slice for every updated frontend with the updated pod backends.

joestringer replied:
(for the reference the comment in my original feedback which is quoted above is the topic of the thread below: https://github.com/cilium/cilium/pull/13362/files/c0a9634f8e410df5119d7cccb14dae981cc29b21..f52c83394229c9cb01ca44cc68387cf5597344f9#r499926266)

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -672,22 +661,53 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
+			}
+		} else {
+			// Collect the unchanged backends for other pods.
+			newFePods = append(newFePods, be)
```
I'm a bit confused why the caller doesn't just pass down the frontend and all desired backends, why do we need to fold in the previous `frontendMapping.podBackends` here? The function is called `upsertFrontendMapping()`, ie "update or insert this set", not "merge" or "append".

aditighag replied:
We don't compute delta for pod backend updates. Whenever frontend's pod backends change, we will override the existing service entry with the updated backends. The "upsert" name looks fine to me, if there are currently no backends for a frontend, then it'll insert the new pod backends, else it'll update the backends slice with the pod backends. Also, keep in mind that we need to maintain the frontend to backends mapping (e.g., fe tcp-> be tcp, fe udp->be udp).

joestringer replied:
My thought on this was not fully developed when I previously asked the question, so bear with me. Here's my understanding of the overall logic going on here:

* `processConfigWithSinglePort()` / `processConfigWithNamedPorts()` loop through the pods
  * For each pod, it collects the backends and calls `upsertFrontendMapping()`
    * This logic (highlighted by this comment thread) folds the new backends into the current `frontendMapping` then calls `rpm.UpsertService()` to call down to the datapath

Why not have the following:
* `processConfigWithSinglePort()` / `processConfigWithNamedPorts()` loop through the pods to collect the backends
* Then invoke `upsertFrontendMapping()` once (through to `rpm.UpsertService()` once) with all backends

Would this allow us to simplify the logic in this function?

aditighag replied:
>Also, keep in mind that we need to maintain the frontend to backends mapping (e.g., fe tcp-> be tcp, fe udp->be udp).

We need to call `rpm.upsertFrontendMapping` and `rpm.upsertService` for every frontend. If you feel we are still not on the same page, I'm happy to discuss it offline in more details.

joestringer replied:
I agree that we need to call both of these for every frontend, my comment above is referring to backends (the inner loop in `processConfigWithNamedPorts()`), not the frontends (the outer loop in `processConfigWithNamedPorts()`).

aditighag replied:
processConfigWith* processes updates for a given pod, and then we aggregate all the backend pods for a frontend with `upsertFrontendMapping`. I'm not sure which part can be simplified here.

joestringer replied:
I guess I just don't understand why we can't aggregate the backend pods in processConfigWith* into a simple slice and call `upsertFrontendMapping()` once with all backends at once.

As far as I understand at the moment the code does:
* Iterate through all frontends (processConfigWith*)
  * Iterate through all backends (processConfigWith*)
    * Iterate through all previously-iterated backends (`upsertFrontendMapping()`)

If processConfigWith* collected all backends and invoked `upsertFrontendMapping()` once per frontend, then the third loop could be removed, and furthermore we would avoid issuing a `upsertService()` once per backend and instead issue once per frontend.

aditighag replied:
The data structures that you suggested might not be amenable to that sort of a change.

Currently, `upsertFrontendMapping` receives only new pod backends, which are then merged with the other backends of the frontend. If we aggregate all the pod backend updates in a slice, we don't know which pods have been updated, and we would still need a 2nd inner loop.

* Iterate through all frontends (processConfigWith*)
   * Iterate through all backends (processConfigWith*)
       * Collect pod backend updates in a slice
   * Iterate through all frontend backends
      * Iterate through the pod backend updates slice


I'll need to introduce some other data structures and maintain additional state in there. 
This could work - 

```
type podBeUpdate {
    bes []backend
    updatePodBe bool
}
```
podUpdatesMap := make(map[podID][]podBeUpdate)
* iterate through all pod backends
    * Add backends to the map

* Iterate through all frontend backends
    * if be.podID is in the map
       * Override the backends for this podID (flip the updatePodBe flag to false so that subsequent entries can be skipped )
    * else
       * Add the unchanged pod be

Iterate all the entries with updatePodBe flag still set to true (these are new pods)

Override frontend backends

This is mostly what we are doing in `upsertFrontendMapping`, but with more simplicity. IMO, with the current scale, I don't see much value in adding the additional complexity at this time.

aditighag replied:
As a middle ground, I can move `upsertService` out of `upsertFrontendMapping` so that it's called just once for every frontend.

aditighag replied:
Done. PTAL.

joestringer replied:
> As a middle ground, I can move upsertService out of upsertFrontendMapping so that it's called just once for every frontend.

That's what I was trying (but apparently failing) to communicate :sweat_smile: 

When I started this thread, I thought that the backend management (upper portion) in this function + the upsertService only made sense to run once per frontend, but I didn't really understand what the `rpm.policyPods` management portion did. Having dug a little deeper, I can see that the policyPods mapping update portion needs to be invoked for every pod that is touched by the policy. That seems to be the piece that I was failing to understand above. FWIW I toyed with the code and here's the incremental diff that achieves what I had in mind:

```
diff --git a/pkg/redirectpolicy/manager.go b/pkg/redirectpolicy/manager.go
index e83bfd19f87d..e1fbbd1929c8 100644
--- a/pkg/redirectpolicy/manager.go
+++ b/pkg/redirectpolicy/manager.go
@@ -75,7 +75,7 @@ type Manager struct {
        // Stores mapping of redirect policy serviceID to the corresponding policyID for
        // easy lookup in policyConfigs
        policyServices map[k8s.ServiceID]policyID
-       // Stores mapping of pods to redirect policies that select the pods
+       // Stores desired mapping of pods to redirect policies that select the pods
        policyPods map[podID][]policyID
        // Stores redirect policy configs indexed by policyID
        policyConfigs map[policyID]*LRPConfig
@@ -615,13 +615,14 @@ func (rpm *Manager) processConfigWithSinglePort(config *LRPConfig, pods ...*podM
                                        bes6 = append(bes6, be)
                                }
                        }
-               }
-               if len(bes4) > 0 {
-                       rpm.upsertFrontendMapping(config, feM, pod.id, bes4)
-               } else if len(bes6) > 0 {
-                       rpm.upsertFrontendMapping(config, feM, pod.id, bes6)
+                       rpm.updatePodPolicyMapping(config.id, pod.id)
                }
        }
+       if len(bes4) > 0 {
+               rpm.upsertFrontendMapping(config, feM, bes4)
+       } else if len(bes6) > 0 {
+               rpm.upsertFrontendMapping(config, feM, bes6)
+       }
        return
 }

@@ -675,59 +676,43 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
                                                        bes6 = append(bes6, be)
                                                }
                                        }
+                                       rpm.updatePodPolicyMapping(config.id, pod.id)
                                }
                        }
-                       if len(bes4) > 0 {
-                               rpm.upsertFrontendMapping(config, feM, pod.id, bes4)
-                       } else if len(bes6) > 0 {
-                               rpm.upsertFrontendMapping(config, feM, pod.id, bes6)
-                       }
                }
-       }
-}
-
-// upsertFrontendMapping updates policy config internal state and upserts
-// service with the policy frontend and mapped backends.
-func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
-       newFePods := make([]backend, 0, len(frontendMapping.podBackends)+len(backends))
-       updatePodBes := true
-       // Update the frontend mapped backends slice, keeping the order same.
-       for _, be := range frontendMapping.podBackends {
-               if be.podID == podID {
-                       if updatePodBes {
-                               updatePodBes = false
-                               // Get the updated backends for the given pod.
-                               newFePods = append(newFePods, backends...)
-                       }
-               } else {
-                       // Collect the unchanged backends for other pods.
-                       newFePods = append(newFePods, be)
+               if len(bes4) > 0 {
+                       rpm.upsertFrontendMapping(config, feM, bes4)
+               } else if len(bes6) > 0 {
+                       rpm.upsertFrontendMapping(config, feM, bes6)
                }
        }
-       if updatePodBes {
-               // New backend pod for the frontend
-               newFePods = append(newFePods, backends...)
-       }
-       frontendMapping.podBackends = newFePods
+}

+func (rpm *Manager) updatePodPolicyMapping(policy policyID, podID podID) {
        if podPolicies, ok := rpm.policyPods[podID]; ok {
                newPodPolicy := true
                for _, poID := range podPolicies {
                        // Existing pod policy update
-                       if poID == config.id {
+                       if poID == policy {
                                newPodPolicy = false
                                break
                        }
                }
                if newPodPolicy {
                        // Pod selected by a new policy
-                       rpm.policyPods[podID] = append(rpm.policyPods[podID], config.id)
+                       rpm.policyPods[podID] = append(rpm.policyPods[podID], policy)
                }
        } else {
                // Pod selected by a policy for the first time
-               pp := []policyID{config.id}
+               pp := []policyID{policy}
                rpm.policyPods[podID] = pp
        }
+}
+
+// upsertFrontendMapping updates policy config internal state and upserts
+// service with the policy frontend and mapped backends.
+func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, backends []backend) {
+       frontendMapping.podBackends = backends
        rpm.upsertService(config, frontendMapping)

 ```

It's a little bit simpler (no more backends merging logic) and it now only issues an `UpsertService()` once per frontend. Though when I try it out, I notice some unit test failures, looks like the backends are being added to `frontendMapping` too many times.

aditighag replied:
>It's a little bit simpler (no more backends merging logic) and it now only issues an UpsertService() once per frontend. Though when I try it out, I notice some unit test failures, looks like the backends are being added to frontendMapping too many times.

Yay, looks like the unit tests did their job. :) Anyway the latest commit that I addressed this, and the `upsertService` is being called only once per frontend.

joestringer reviewed (changes requested):
If I'm reading it right, if the pod has multiple backends then we'll issue multiple service updates but we should be able to just calculate all backends for a service at once and then update for all known backends.

aanm reviewed (approved):
Nothing else besides what Joe point it out.

christarazi commented on pkg/service/service.go:
```diff
@@ -544,11 +544,19 @@ func (s *Service) createSVCInfoIfNotExist(p *lb.SVC) (*svcInfo, bool, bool,
+		// Local-redirect service can only override clusterIP service type or itself.
+		if p.Type == lb.SVCTypeLocalRedirect &&
+			(svc.svcType != lb.SVCTypeClusterIP && svc.svcType != lb.SVCTypeLocalRedirect) {
+			err := fmt.Errorf("skip local-redirect service for "+
+				"frontend %v as it overlaps with svc %v of type %v",
+				p.Frontend, svc.svcName, svc.svcType)
```
Nit: the error msg should state what's wrong. In this case, the fact that the service overlapped with another, is the problem. Mentioning the action to be taken, `skipping`, is irrelevant in an error. That should be reserved for a log msg.

```suggestion
			err := fmt.Errorf("local-redirect service overlaps with %v of type %v", svc.svcName, svc.svcType)
```

Something like that.

aditighag replied:
If you expand the commit description for the first commit, you will see the summary of changes. The changes are simple enough that I don't think it's necessary to further split them out.

christarazi replied:
In general, we tend to make renames and other cosmetic changes as separate commits, as Joe pointed out. In this case, the changes are straightforward and small enough. However, the commit title as "Address leftover nits" is not helpful when reviewing the git log in the future. We should express what changed in the title and then why in the body. Could we change it, so that we have a git log with digestible pieces of context into what changed?

aditighag replied:
done

aditighag replied:
> Nit: the error msg should state what's wrong. In this case, the fact that the service overlapped with another, is the problem. Mentioning the action to be taken, `skipping`, is irrelevant in an error. That should be reserved for a log msg.
> 
> Something like that.

TMS. I'll probably also add structured logging here.

aditighag replied:
I thought about this a bit more. The cilium-agent service module should ideally notify the caller (LRP manager/CLI) that it's going to skip the service update because of the overlap so the action _is_ relevant here, and should be included in the error message? 
@brb Any comments?

christarazi replied:
In general, if the caller cares specifically about the error, then defining the error as a sentinel type [1] is the pattern, so that the caller can do `errors.Is(err, ErrOverlappingService)`, for example.

[1]:

```golang
type ErrOverlappingService struct {
    svc string
    type string
}
func (e ErrOverlappingService) Error() string { return fmt.Sprintf("....%s ... %v", e.svc, e.type) }
```

This is just a suggestion. I'm not too familiar with this code to suggest whether we should do this, but if it is a common enough error, then the caller can do the `errors.Is` and then log / alert the user that _it_ will skip the update, etc.

If that doesn't quite make sense in this case, and if it's the callee's responsibility should signal to the caller that it's skipping the update, then feel free to ignore. But just wanted to share how this pattern can be useful in keeping error msgs short and to the point, and keeping the responsibilities of the caller / callee clear. :smile:

aditighag replied:
Sure. In this case, the action to skip a service update is taken by the service module so it makes sense to pass that information to the caller.

christarazi reviewed (changes requested):
One nit for me. Agree with other reviews so far, the first commit should contain more context / be split out into different logical changes.

aditighag committed:
pkg/redirectpolicy Optimize protocol check

Check frontend and backend protocols outside
of forming multiple pod <ip*, port, protocol>
backends, where port/protocol remains same.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

brb commented on pkg/k8s/watchers/cilium_local_redirect_policy.go:
```diff
@@ -72,7 +72,7 @@ func (k *K8sWatcher) ciliumLocalRedirectPolicyInit(ciliumLRPClient *k8s.K8sCiliu
 
 	go lrpController.Run(wait.NeverStop)
-	k.k8sAPIGroups.addAPI(k8sAPIGroupCiliumClusterwideNetworkPolicyV2)
+	k.k8sAPIGroups.addAPI(k8sAPIGroupCiliumLocalRedirectPolicyV2)
```
Q: how did you test that an LRP svc doesn't get removed?

aditighag replied:
Restarted cilium-agent a few times while LRP was still present, and monitored that this logs doesn't show up - "Deleting no longer present service"

brb reviewed (approved):
Checked only https://github.com/cilium/cilium/pull/13362/commits/2002c374b91c5883bcb807908a7213e0abbe2fb9 and https://github.com/cilium/cilium/pull/13362/commits/bc15d06fd18c17e8577b1b3427758b953ad47c56. Both are LGTM.

joestringer commented on pkg/redirectpolicy/manager_test.go:
```diff
@@ -0,0 +1,432 @@
+
+	added, err := m.rpm.AddRedirectPolicy(configAddrType, &fakePodStore{})
+
+	c.Assert(added, Equals, true)
+	c.Assert(err, IsNil)
+	c.Assert(len(m.rpm.policyConfigs), Equals, 1)
+	c.Assert(m.rpm.policyConfigs[configAddrType.id].id.Name, Equals, configAddrType.id.Name)
+	c.Assert(m.rpm.policyConfigs[configAddrType.id].id.Namespace, Equals, configAddrType.id.Namespace)
+	c.Assert(len(m.rpm.policyFrontendsByHash), Equals, 1)
+	c.Assert(m.rpm.policyFrontendsByHash[configAddrType.frontendMappings[0].feAddr.Hash()],
+		Equals, configAddrType.id)
+	c.Assert(len(configAddrType.frontendMappings[0].podBackends), Equals, 2)
+	for i := range configAddrType.frontendMappings[0].podBackends {
+		c.Assert(configAddrType.frontendMappings[0].podBackends[i], checker.Equals, expectedbes[i])
+	}
+	c.Assert(len(m.rpm.policyPods), Equals, 1)
+	c.Assert(len(m.rpm.policyPods[pod1ID]), Equals, 1)
+	c.Assert(m.rpm.policyPods[pod1ID][0], Equals, configAddrType.id)
```
I'm glad to see unit tests, though I would prefer if we took a more black-box approach so that the unit tests aren't so tied into the internal implementation but rather to the interfaces that the object interacts with.

What I mean is, rather than doing:

* AddRedirectPolicy
* Check internal map1 is updated
* Check internal map2 is updated
* Check other internal state
* ...

Instead:

* AddRedirectPolicy
* Check that the podStore, servicemanager get the correct number of lookups/updates with the correct contents.
* Add another pod
* Check that the servicemanager gets frontend updates that reflect the pod change
* ...

With this approach, in future we can modify the internal representation of the data to suit a different implementation, and the unit tests will still work and won't need changing when we change the implementation.

Given that we've just merged this implementation and probably won't go and change it drastically any time soon, I'm not *so* worried to address this in this PR, but this would be good to keep in mind in the future. The reason I bring this up is that there has been other test code written in Cilium in the past which is directly tied to the internals like this, introspecting internal maps and whatnot. When we go to refactor that code, it can be painful to go through each test one-by-one to update the references to refer to the new implementation.

aditighag replied:
I agree with checking the number of calls. In fact, while testing the last fix, I found an issue that wasn't flagged by the unit tests, and made a mental note about this. I'll enhance the unit tests approach.

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -668,42 +679,81 @@ func (rpm *Manager) upsertConfigWithNamedPorts(config *LRPConfig, pods ...*podMe
 			} else if len(bes6) > 0 {
-				rpm.upsertServiceWithBackends(config, feM, pod.id, bes6)
+				rpm.updateFrontendMapping(config, feM, pod.id, bes6)
+				upsertFes = append(upsertFes, feM)
```
This still adds the same `feM` to `upsertFes` once per pod backend right?

aditighag replied:
Good catch! Fixed.

joestringer reviewed (commented):
Minor comment on the tests, non-blocking but something to keep in mind.

The UpsertService change looks correct for `processConfigWithSinglePort()` but it doesn't look like it makes any change for `processConfigWithNamedPorts()`. I think if we resolve that we're good to merge.
